### PR TITLE
Refactor kinesis exporter to be aligned with Kafka Producer behaviour

### PR DESCRIPTION
**Pull request by MovieStoreGuy**
**Description:**
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.>
Updating the kinesis exporter to aligned with the behaviour of the Kafka exporter.

This removes the third party Kinesis Producer Library and  allows for the internal exporter pipeline to allow for things like retries and better reporting of data that is unable to be sent to Kinesis.

**Link to tracking Issue:** Currently a draft, will link the issue when it is ready.

**Testing:** Unit testing has been adding throughout where code has been updated.

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: exporter/awskinesisexporter/factory.go [hunk 1]
```diff
@@ -45,59 +53,49 @@ func createDefaultConfig() config.Exporter {
...
+		cfgs = append(cfgs, &aws.Config{Endpoint: aws.String(c.AWS.KinesisEndpoint)})
+	}
+
+	producer, err := producer.NewBatcher(kinesis.New(sess, cfgs...), c.AWS.StreamName,
+		producer.WithLogger(params.Logger),
+	)
+
+	return &Exporter{
+		producer: producer,
+		batcher:  batch.NewJaegerBatcher(),
```
Comment (MovieStoreGuy) [inline exporter/awskinesisexporter/factory.go:99]: Needs to accept batch size and record size.

File: exporter/awskinesisexporter/internal/batch/batch.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+// AddProtobufV2 accepts the protobuf message and marshal it into a record
+// ready to submit to kinesis.
+func (b *Batch) AddProtobufV2(message protov2.Message, key string) error {
+	data, err := protov2.Marshal(message)
+	if err != nil {
+		return err
+	}
+	return b.addRaw(data, key)
+}
+
```
Comment (MovieStoreGuy) [inline exporter/awskinesisexporter/internal/batch/batch.go:115]: Potentially a need to add for `AddJson(value interface{}, key string)`, unsure for this moment. Will leave for a later PR if that is the case.

File: exporter/awskinesisexporter/internal/batch/batch.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+// to be written to he kinesis.PutRecords endpoint
+func (b *Batch) Chunk() [][]*kinesis.PutRecordsRequestEntry {
+	length, chunk := len(b.records), make([][]*kinesis.PutRecordsRequestEntry, 0, len(b.records)/b.maxBatchSize+1)
+	for i, end := 0, min(length, b.maxBatchSize); i < length; i, end = i+b.maxBatchSize, i+end+min(b.maxBatchSize, length-end) {
+		chunk = append(chunk, b.records[i:end])
+	}
+	return chunk
+}
+
+func min(a, b int) int {
```
Comment (MovieStoreGuy) [inline exporter/awskinesisexporter/internal/batch/batch.go:126]: Can not wait for generics to remove crap like this.

File: exporter/awskinesisexporter/internal/batch/encode_jaeger.go [hunk 4]
```diff
@@ -12,47 +12,49 @@
...
 	"go.opentelemetry.io/collector/consumer/consumererror"
 	"go.opentelemetry.io/collector/model/pdata"
 
 	jaegertranslator "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/jaeger"
 )
 
-type jaeger struct {
-	kinesis *awskinesis.Exporter
-}
+type jaeger struct{}
```
Comment (MovieStoreGuy) [inline exporter/awskinesisexporter/internal/batch/encode_jaeger.go:24]: Need to update to store max allowed batch size and record size when creating a batch to forward to kinesis.

File: exporter/awskinesisexporter/factory.go [hunk 5]
```diff
@@ -45,59 +53,49 @@ func createDefaultConfig() config.Exporter {
...
+
+func NewMetricsExporter(ctx context.Context, params component.ExporterCreateSettings, conf config.Exporter) (component.MetricsExporter, error) {
+	return createExporter(ctx, params, conf)
+}
+
+func NewLogsExporter(ctx context.Context, params component.ExporterCreateSettings, conf config.Exporter) (component.LogsExporter, error) {
+	return createExporter(ctx, params, conf)
+}
+
+func createExporter(_ context.Context, params component.ExporterCreateSettings, conf config.Exporter) (*Exporter, error) {
```
Comment (irisgve) [inline exporter/awskinesisexporter/factory.go:73]: To align this better with the Kafka exporter, should we consider using more of exporterhelper methods? i.e. for instantiating a new exporter we use `exporterhelper.NewXXXExporter`

Comment (MovieStoreGuy) [reply exporter/awskinesisexporter/factory.go:73]: I am in for that, I need to look into detail what methods they are using and I can carry them over here.

**Comment by MovieStoreGuy** [comment]
My git habits got in the way so I have squashed my local and pushed it up here. Have also rebased with main so I am hoping the issue with linting is resolved.

File: exporter/awskinesisexporter/factory.go (hunk 5 above)
Comment (irisgve) [reply exporter/awskinesisexporter/factory.go:72]: iirc it always interfacing with retry and queueing configurations so defo worth checking out. Creating via the exporterhelper also enables the internal metrics for the exporter by default I believe

File: exporter/awskinesisexporter/config.go [hunk 6]
```diff
@@ -21,35 +21,16 @@ import (
 // AWSConfig contains AWS specific configuration such as awskinesis stream, region, etc.
```
Comment (irisgve) [inline exporter/awskinesisexporter/config.go:21]: Minor: let's add these configs in README i.e. https://github.com/atlassian-forks/opentelemetry-collector-contrib/tree/master/exporter/awskinesisexporter

File: exporter/awskinesisexporter/README.md [hunk 7]
```diff
@@ -1,3 +1,43 @@
 # Kinesis Exporter
 
-To be added.
\ No newline at end of file
+
+The kinesis exporter currently exports Jaeger traces (dynamic encodings coming) to the configured kinesis stream.
```
Comment (irisgve) [inline exporter/awskinesisexporter/README.md:4]: Nice thanks for this

File: exporter/awskinesisexporter/exporter.go [hunk 8]
```diff
@@ -16,32 +16,68 @@ package awskinesisexporter
...
 }
 
-var _ component.TracesExporter = (*Exporter)(nil)
+var (
+	_ component.TracesExporter  = (*Exporter)(nil)
+	_ component.MetricsExporter = (*Exporter)(nil)
+	_ component.LogsExporter    = (*Exporter)(nil)
+)
+
+func createExporter(c config.Exporter, log *zap.Logger) (*Exporter, error) {
```
Comment (irisgve) [inline exporter/awskinesisexporter/exporter.go:47]: This by default creates a trace exporter given that the batching uses `NewJaeger`. What's the plan to extend with the other types and does it make more sense to name it like so and create different exporters for each.

**Pull request closed without merging**